extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elasticsearch/impl/ElasticProcessorMetaSupplier.java

```diff
@@ -0,0 +1,210 @@
+                if (thisNodeCandidates.isEmpty()) {
+                    continue;
+                }
+                Shard shard = thisNodeCandidates.remove(0);
+
+                List<Shard> nodeShards = nodeAssigned.computeIfAbsent(address, (key) -> new ArrayList<>());
+                nodeShards.add(shard);
+
+                nodeCandidates.values().forEach(candidates ->
+                        candidates.removeIf(next -> next.indexShard().equals(shard.indexShard())));
+
+                assignedShards.add(shard.indexShard());
+            }
+        }
+        if (assignedShards.size() != uniqueShards) {
+            throw new JetException("Not all shards have been assigned assigned");
+        }
+        return nodeAssigned;
+    }
+
+
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
+    private List<Shard> readShards() throws IOException {
+        try (RestHighLevelClient client = builder.clientSupplier().get()) {
+            SearchRequest sr = builder.searchRequestSupplier().get();
+            Request r = new Request("GET", "/_cat/shards/" + String.join(",", sr.indices()));
+            r.addParameter("format", "json");
+            Response res = client.getLowLevelClient().performRequest(r);
+
+            try (InputStreamReader reader = new InputStreamReader(res.getEntity().getContent(), UTF_8)) {
+                JsonArray array = Json.parse(reader).asArray();
+                List<Shard> shards = new ArrayList<>(array.size());
+                for (JsonValue value : array) {
+                    Optional<Shard> shard = convertToShard(value);
+                    shard.ifPresent(shards::add);
+                }
+
+                if (logger.isFineEnabled()) {
+                    logger.log(FINER, "Shards " + shards);
+                }
+                return shards;
+            }
+        }
+    }
+
+    private Optional<Shard> convertToShard(JsonValue value) {
+        JsonObject object = value.asObject();
+        // TODO IndexShardState.STARTED but this is deeply inside elastic, should we mirror the enum?
+        if ("STARTED".equals(object.get("state").asString())) {
+            Shard shard = new Shard(
+                    object.get("index").asString(),
+                    Integer.parseInt(object.get("shard").asString()),
+                    Prirep.valueOf(object.get("prirep").asString()),
+                    Integer.parseInt(object.get("docs").asString()),
+                    object.get("state").asString(),
+                    object.get("ip").asString(),
+                    object.get("node").asString()
+            );
+            return Optional.of(shard);
+        } else {
+            return empty();
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Function<? super Address, ? extends ProcessorSupplier> get(@Nonnull List<Address> addresses) {
+        if (builder.slicing()) {
```

Comment:
What happens if builder has `slicing=true` and `coLocatedReading=true` ?

Reply:
Wanted to ignore / forbid this combination originally, but it might make sense in certain situations.

See the updates.